replit_river/codegen/client.py

```diff
@@ -49,6 +53,59 @@ def encode_type(
     if isinstance(type, RiverNotType):
         return ("None", ())
     if isinstance(type, RiverUnionType):
+        # First check if it's a discriminated union.
+        if len(type.anyOf) > 0 and all(
+            isinstance(t, RiverConcreteType)
+            and t.type == "object"
+            and t.properties
+            and (not t.patternProperties or "^(.*)$" not in t.patternProperties)
+            for t in type.anyOf
+        ):
+            # We have established that it is a union-of-objects. Now let's see if
+            # there is a discriminator field common among all options.
+            literal_fields = set[str]()
+            for i, t in enumerate(type.anyOf):
+                # Mypy already forgot this fact D:
```

Comment:
since it's a `List` you're always gonna bubble back up to the ascribed type, unfortunately. Maybe just break out an explicit binding with the known type, if you don't want to keep doing this dance.

Reply:
yeah, that's a good idea. did that and added a big comment about it.